include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -0,0 +1,184 @@
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan) {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
+        }
+        return scan;
+    }
+};
+
+template<typename T, class Binop, class SharedAccessor>
+struct WorkgroupScanTail
+{
+    bool isExclusive;
+    T identity;
+    T firstLevelScan;
+    uint lastInvocation;
+    uint scanStoreIndex;
+
+    static WorkgroupScanTail create(bool isExclusive, T identity, T firstLevelScan, uint lastInvocation, uint scanStoreIndex)
+    {
+        WorkgroupScanTail<T, Binop, SharedAccessor> wst;
+        wst.isExclusive = isExclusive;
+        wst.identity = identity;
+        wst.firstLevelScan = firstLevelScan;
+        wst.lastInvocation = lastInvocation;
+        wst.scanStoreIndex = scanStoreIndex;
+        return wst;
+    }
+
+    T operator()()
+    {
+        Binop binop;
+        SharedAccessor sharedAccessor;
+        sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+        
+        const uint subgroupId = glsl::gl_SubgroupID();
+        if(lastInvocation >= glsl::gl_SubgroupSize())
+        {
+            uint scanLoadIndex = scanStoreIndex + glsl::gl_SubgroupSize();
+            const uint shiftedInvocationIndex = gl_LocalInvocationIndex + glsl::gl_SubgroupSize();
+            const uint currentToHighLevel = subgroupId - shiftedInvocationIndex;
+            for(uint logShift = (firstbithigh(lastInvocation) / glsl::gl_SubgroupSizeLog2() - 1u) * glsl::gl_SubgroupSizeLog2(); logShift > 0u; logShift -= glsl::gl_SubgroupSizeLog2())
+            {
+                uint lastInvocationInLevel = lastInvocation >> logShift;
+                sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+                const uint currentLevelIndex = scanLoadIndex - (lastInvocationInLevel + 1u);
+                if(shiftedInvocationIndex <= lastInvocationInLevel)
+                {
+                    sharedAccessor.main.set(currentLevelIndex, binop(sharedAccessor.main.get(scanLoadIndex+currentToHighLevel), sharedAccessor.main.get(currentLevelIndex)));
+                }
+                scanLoadIndex = currentLevelIndex;
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            
+            if(gl_LocalInvocationIndex <= lastInvocation && subgroupId != 0u)
+            {
+                const uint higherLevelExclusive = sharedAccessor.main.get(subgroupId - 1u);
+                firstLevelScan = binop(higherLevelExclusive, firstLevelScan);
+            }
```

Comment:
this needs to be re-examined to be in-line with the Head suggestions

also the code has the same bugs w.r.t smem addressing

Reply:
pseudocode of loop
```
store higher level with all participating invocations into smem to allow broadcast

barrier

increase participating invocation set
read previously stored using subgroupID of participating invocations
step back with the store offset
```